File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/00 - MY_Attempts/merge_sort.py
from typing import List 

class Solution:
    def mergeSort(self, nums, left, right):
        # The mergeSort function is a recursive sorting algorithm that sorts the 'nums' list
        # within the specified range [left, right].
        if left >= right:
            return  # If the range contains 0 or 1 elements, it is already sorted.

        mid = left + (right - left) // 2  # Calculate the middle index of the range.
        self.mergeSort(nums, left, mid)  # Recursively sort the left half of the range.
        self.mergeSort(nums, mid + 1, right)  # Recursively sort the right half of the range.
        self.merge(nums, left, mid, right)  # Merge the sorted left and right halves.

    def merge(self, nums, left, mid, right):
        # The merge function merges two sorted subarrays [left, mid] and [mid+1, right] into a single sorted subarray.

        temp = []  # Temporary list to store the merged result.
        i, j = left, mid + 1  # Initialize pointers for the two subarrays.

        # Compare elements from both subarrays and append the smaller one to the 'temp' list.
        while i <= mid and j <= right:
            if nums[i] < nums[j]:
                temp.append(nums[i])
                i += 1
            else:
                temp.append(nums[j])
                j += 1

        # If there are remaining elements in the left subarray, append them to 'temp'.
        while i <= mid:
            temp.append(nums[i])
            i += 1

        # If there are remaining elements in the right subarray, append them to 'temp'.
        while j <= right:
            temp.append(nums[j])
            j += 1

        # Copy the merged elements from 'temp' back to the original 'nums' list.
        nums[left:right + 1] = temp

    def sortArray(self, nums: List[int]) -> List[int]:
        # The sortArray function is the entry point for the merge sort algorithm.
        self.mergeSort(nums, 0, len(nums) - 1)  # Start the merge sort with the full range.
        return nums  # Return the sorted 'nums' list.


nums = [5, 2, 3, 1]
print("NUMS:", nums)
solution = Solution()  # Create an instance of the Solution class
sorted_nums = solution.sortArray(nums)
print("SORTED:", sorted_nums)

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/06-Sorting/CppEx/MergeSort.cpp
#include <stdio.h>
#include <vector>
#include <iostream>

using namespace std;

///-----------------------------------------------------
/// MergeSort: O(NlogN) with O(N) extra space
/// 
void MS(vector<int>& A, int left, int right, vector<int>& B) {
	if (left >= right) return;  // A list of size <= 1 is already sorted

	int middle = left + (right-left) / 2;
	MS(A, left, middle, B);        // Sort the LHS of size N/2
	MS(A, middle + 1, right, B);   // Sort the RHS of size N/2

	// Merge the two halves: O(N)
	int i = left;
	int j = middle + 1;
	int k = 0;
	while (i <= middle && j <= right) {
		if (A[i] <= A[j]) B[k++] = A[i++];
		else              B[k++] = A[j++];
	} //end-while

	// Copy the remaining items from A to B
	while (i <= middle) B[k++] = A[i++];
	while (j <= right) B[k++] = A[j++];

	// Copy the sorted items from B to A
	k = left;
	i = 0;
	while (i < right - left + 1) A[k++] = B[i++];
} //end-MS

void MergeSort(vector<int>& A) {
	if (A.size() <= 1) return;

	vector<int> B(A.size());   // Allocate a temporary buffer of size N
	MS(A, 0, (int)A.size()-1, B);
} //end-MergeSort


int main(){
	vector<int> nums = { 5, 10, 3, 2, 8, 5, 12, 20, 25, 17, 5, 1, 9, 3, 13, 12, 1 };

	printf("Initial sequence:\n");
	for (auto num: nums) cout << num << ", ";
	cout << "\b\b \n";

	MergeSort(nums);

	printf("Sorted sequence:\n");
	for (auto num: nums) cout << num << ", ";
	cout << "\b\b \n";

	return 0;
} // end-main

